### ISSUE TITLE
Please optimize loading z-sy-h (time to first prompt in a new shell)

### ISSUE THREAD

**Issue by tmpm697** [zsh-users/zsh-syntax-highlighting#734]
when test this plugin, I've notice that this function `_zsh_highlight_load_highlighters` always on top of function that consumes resources that slow down prompt a bit.

Is there any reason why this function consumes so much resources? even higher than compinit.

`for i in $(seq 1 10); do time zsh -i -c exit; done` with `zmodload zsh/zprof`

**Comment by danielshahaf** [comment]
Clarify, please: Are you commenting about the initial loading stage (the time between opening the terminal emulator and the _first_ prompt appearing) or about the code that runs upon every  keystroke?

**Comment by danielshahaf** [comment]
What highlighters do you have enabled? What are the version numbers of zsh and z-sy-h? What's the zprof output?

**Comment by danielshahaf** [comment]
Here's my data:

```zsh
% head .zshenv .zshrc
==> .zshenv <==
zmodload zsh/zprof

==> .zshrc <==
zprof -c
autoload compinit
compinit
source /usr/share/zsh-syntax-highlighting/zsh-syntax-highlighting.zsh
zprof
% 
```

```
% repeat 3 >/dev/null ZDOTDIR=$PWD zsh -ic exit 
%                     ZDOTDIR=$PWD zsh -ic exit 
num  calls                time                       self            name
-----------------------------------------------------------------------------------
 1)    2          10.90     5.45   42.87%     10.90     5.45   42.87%  compaudit
 2)    1          16.94    16.94   66.66%      6.05     6.05   23.79%  compinit
 3)    1           3.97     3.97   15.62%      3.97     3.97   15.62%  _zsh_highlight_bind_widgets
 4)    1           4.09     4.09   16.08%      3.83     3.83   15.05%  _zsh_highlight_load_highlighters
 5)    1           0.36     0.36    1.40%      0.36     0.36    1.40%  is-at-least
 6)    2           0.32     0.16    1.26%      0.32     0.16    1.26%  add-zsh-hook

-----------------------------------------------------------------------------------
⋮
```

The ratio of compinit to z-sy-h is about 2:1.  Of the time spent in z-sy-h, the widget binding half will likely diminish when the feature/redrawhook branch is merged.  As to the other half, 91% of the work in `_zsh_highlight_load_highlighters` hides in the single statement `. "$highlighter_dir${highlighter}-highlighter.zsh"` [confirmed by dropping an anonymous function around that statement], so once we fix #252 _that_ will be addressed too, thanks to `zcompile`.

I have labelled #252 with the `Performance` label.  Thanks.

Other than that… Well, I guess we should keep this issue open, and close it once redrawhook is merged and #252 is fixed, even though there'll likely be nothing to do here once those two are merged.

@tmpm697 Did I miss anything?

**Comment by tmpm697** [comment]
> the time between opening the terminal emulator and the first prompt appearing

Yes, it is.

I have pretty much stable output as this:
```
num  calls                time                       self            name
-----------------------------------------------------------------------------------
 1)    6          12.45     2.07   37.23%      8.04     1.34   24.03%  (anon)
 2)    2           6.79     3.40   20.31%      6.79     3.40   20.31%  _zsh_highlight_bind_widgets
 3)    1           5.42     5.42   16.20%      5.32     5.32   15.92%  _zsh_highlight_load_highlighters
 4)    1           4.44     4.44   13.29%      4.44     4.44   13.29%  enable-fzf-tab
 5)    1           3.65     3.65   10.92%      3.65     3.65   10.92%  compinit
 6)    6           1.42     0.24    4.24%      1.42     0.24    4.24%  is-at-least
 7)    5           1.26     0.25    3.78%      1.26     0.25    3.78%  add-zsh-hook
 8)    1           1.52     1.52    4.55%      0.49     0.49    1.46%  prompt_powerlevel9k_teardown
 9)   12           0.44     0.04    1.32%      0.44     0.04    1.32%  _fzf_tab_add_default
10)    1           2.12     2.12    6.34%      0.41     0.41    1.21%  prompt_powerlevel9k_setup
11)    1           4.00     4.00   11.96%      0.35     0.35    1.04%  _zcompinit_custom
12)    1           0.77     0.77    2.32%      0.24     0.24    0.72%  _p9k_preinit
13)    1           0.54     0.54    1.60%      0.23     0.23    0.68%  gitstatus_start_p9k_
14)    2           0.13     0.07    0.39%      0.13     0.07    0.39%  _p9k_restore_special_params
15)    1           0.12     0.12    0.35%      0.12     0.12    0.35%  p10k
16)    1           1.63     1.63    4.88%      0.11     0.11    0.33%  _p9k_setup
```

I do compinit each 24h so it does not took much resource but `zsh-syntax-highlighting` still. Beside that I can't know `(anon)` is from `zsh-syntax-highlighting` or not.

For sure if I disable `zsh-syntax-highlighting`, something else would took up `%` but speed would quite good improved.

**Comment by danielshahaf** [comment]
> num  calls                time                       self            name
> -----------------------------------------------------------------------------------
>  1)    6          12.45     2.07   37.23%      8.04     1.34   24.03%  (anon)
>  2)    2           6.79     3.40   20.31%      6.79     3.40   20.31%  _zsh_highlight_bind_widgets
>  3)    1           5.42     5.42   16.20%      5.32     5.32   15.92%  _zsh_highlight_load_highlighters

_zsh_highlight_bind_widgets was called twice but _zsh_highlight_load_highlighters was called once.

There is no codepath in z-sy-h that calls _zsh_highlight_bind_widgets twice.

I'm guessing that one of your other plugins calls that function too.  You'll have to take those 3.40 seconds of slowdown up with the maintainer of that plugin, whichever it is.

> Beside that I can't know `(anon)` is from `zsh-syntax-highlighting` or not.
> 

"(anon)" is a cumulative entry for _all_ anonymous functions covered by that zprof output.  To see z-sy-h's contribution, disable z-sy-h entirely and compare the figures with and without z-sy-h.

Also, could you drop a line to zsh-users@zsh.org about making zprof emit file:line coordinates from funcstack for anonymous functions?

> For sure if I disable `zsh-syntax-highlighting`, something else would took up `%` but speed would quite good improved.
> 

Mind if I ask what kind of environment this is (CPU, disks, and $ZSH_VERSION and $ZSH_PATCHLEVEL)?  5.32 + (6.79/2) seconds — ignoring the double call, which isn't our fault — for loading z-sy-h is quite a lot.  You don't have to answer.

**Comment by danielshahaf** [comment]
Duplicate of #635.

**Comment by danielshahaf** [comment]
**Note about the issue title:** I've changed this issue's title back to the original one and moved the title "Please optimize loading z-sy-h (time to first prompt in a new shell)" to #635.

**Comment by tmpm697** [comment]
compare to fast-syntax-hight, I think i'll go for it.
```
 1)    7          13.03     1.86   43.53%      8.42     1.20   28.12%  (anon)
 2)    2           8.31     4.15   27.75%      8.31     4.15   27.75%  _zsh_highlight_bind_widgets
 3)    1           4.45     4.45   14.87%      4.45     4.45   14.87%  enable-fzf-tab
 4)    1           3.48     3.48   11.63%      3.48     3.48   11.63%  compinit
 5)    5           1.28     0.26    4.28%      1.28     0.26    4.28%  is-at-least
 6)    4           1.06     0.27    3.56%      1.06     0.27    3.56%  add-zsh-hook
 7)    1           1.89     1.89    6.31%      0.59     0.59    1.98%  _p9k_preinit
 8)   12           0.54     0.05    1.81%      0.54     0.05    1.81%  _fzf_tab_add_default
 9)    1           1.29     1.29    4.33%      0.44     0.44    1.48%  gitstatus_start_p9k_
10)    1           3.87     3.87   12.94%      0.39     0.39    1.31%  _zcompinit_custom
11)    1           1.26     1.26    4.21%      0.35     0.35    1.16%  prompt_powerlevel9k_teardown
12)    1           1.64     1.64    5.47%      0.22     0.22    0.74%  prompt_powerlevel9k_setup
13)    1           0.12     0.12    0.39%      0.12     0.12    0.39%  p10k
14)    1           1.35     1.35    4.52%      0.09     0.09    0.31%  _p9k_setup
15)    2           0.09     0.05    0.31%      0.09     0.05    0.31%  _p9k_restore_special_params
16)    1           0.09     0.09    0.30%      0.09     0.09    0.30%  -fast-highlight-fill-option-variables
```

improved 10% faster than z-sy-h.

**Comment by danielshahaf** [comment]
Thanks for the feedback, @tmpm697.  

To be clear, the only reason I closed this issue is that there's another issue about the very same matter.  I do **not** consider this issue resolved yet.  (You may want to subscribe to #635, if you haven't already.)

Regarding z-sy-h, merging redrawhook is a priority.  Once that's out of the way, #252 should be trivial to implement.  At that point z-sy-h should load much faster than it does now.  However, merging redrawhook requires making a change to zsh itself first (https://github.com/zsh-users/zsh/pull/57; we're working on that now.  (See, for example, the long, ongoing discussion on https://github.com/zsh-users/zsh-autosuggestions/issues/529.)

All that said, we'll respect whichever decision you make.  If you end up choosing f-sy-h, then please give my regards to @psprint :-)
